### ISSUE TITLE
Problem: Zyre announcement issues

### ISSUE THREAD

**Issue by willkelly** [zeromq/zbroker#85]
In our production environment, we are seeing brokers fail to announce their presence to one another appropriately.  In the example, one of the brokers announces and determines that the broker of its name has vanished, and announces again, and vanishes again.  The other side does not see anything at all, though we do see udp traffic on the zyre port in tcpdump.

Error in test "scripts/simple_writer.yml"
### Test script

```
sleep 1
timeout 3000
open pipe1 write
write pipe1 abc
close pipe1 write
```
### Script Log

```
Set prefix to 1
Set broker to 1-0
Sleeping for 1 seconds
Set timeout to 3000
Opening descriptor "1-0|>1-pipe1"
Opened pipe "pipe1" for write
Writing "abc" to pipe "pipe1"
Unexpected exception: timeouterror
Test failed

```
### Broker Log

```
zbroker service/0.0.1
Copyright (c) 2014 the Contributors
This Software is provided under the MPLv2 License on an "as is" basis,
without warranty of any kind, either expressed, implied, or statutory.

14-05-19 15:45:07 I: starting zpipes broker using config in '/tmp/tmp_D7WgI/zbroker.cfg'
14-05-19 15:45:07 I: joining cluster as A56A57
14-05-19 15:45:07 N: starting zpipes_server service
14-05-19 15:45:07 I: ZPIPES server appeared at A4BDB9
14-05-19 15:45:07 I: ZPIPES server appeared at A56A57
14-05-19 15:45:07 I: ZPIPES server vanished from A56A57
14-05-19 15:45:07 I: ZPIPES server appeared at A56A57
14-05-19 15:45:07 I: ZPIPES server vanished from A56A57
14-05-19 15:45:07 I: ZPIPES server appeared at A56A57
14-05-19 15:45:09 D:    139:                    : start:
14-05-19 15:45:09 D:    139:                    :     OUTPUT
14-05-19 15:45:09 D:    139:                    :         $ lookup or create pipe
14-05-19 15:45:09 D:    139:1-pipe1             :         $ open pipe writer
14-05-19 15:45:09 1-pipe1: attach local writer
14-05-19 15:45:09 1-pipe1: broadcast we are now writer
14-05-19 15:45:09 D:    139:1-pipe1             : open local writer
14-05-19 15:45:09 D:    139:1-pipe1             :         > before writing
14-05-19 15:45:09 D:    139:1-pipe1             : before writing:
14-05-19 15:45:09 D:    139:1-pipe1             :     ok
14-05-19 15:45:09 D:    139:1-pipe1             :         $ send OUTPUT_OK
14-05-19 15:45:09 D:    139:1-pipe1             :         > writing
14-05-19 15:45:09 D:    139:1-pipe1             : writing:
14-05-19 15:45:09 D:    139:1-pipe1             :     WRITE
14-05-19 15:45:09 D:    139:1-pipe1             :         $ process write request
14-05-19 15:45:09 D:    139:1-pipe1             :         > processing write
14-05-19 15:45:12 D:    139:1-pipe1             : processing write:
14-05-19 15:45:12 D:    139:1-pipe1             :     wakeup
14-05-19 15:45:12 D:    139:1-pipe1             :         $ send WRITE_TIMEOUT
14-05-19 15:45:12 D:    139:1-pipe1             :         > writing
interrupted
14-05-19 15:45:13 N: terminating zpipes_server service
Context was terminated
Context was terminated
Context was terminated
Context was terminated

```
### Test script

```
sleep 1
timeout 3000
open pipe1 read
read pipe1 3 abc
close pipe1 read
```
### Script Log

```
Set prefix to 1
Set broker to 1-1
Sleeping for 1 seconds
Set timeout to 3000
Opening descriptor "1-1|1-pipe1"
Opened pipe "pipe1" for read
Reading 3 bytes from pipe "pipe1"
Unexpected exception: timeouterror
Test failed

```
### Broker Log

```
zbroker service/0.0.1
Copyright (c) 2014 the Contributors
This Software is provided under the MPLv2 License on an "as is" basis,
without warranty of any kind, either expressed, implied, or statutory.

14-05-19 15:45:07 I: starting zpipes broker using config in '/tmp/tmp9B4wt1/zbroker.cfg'
14-05-19 15:45:07 I: joining cluster as A4BDB9
14-05-19 15:45:07 N: starting zpipes_server service
14-05-19 15:45:09 D:     50:                    : start:
14-05-19 15:45:09 D:     50:                    :     INPUT
14-05-19 15:45:09 D:     50:                    :         $ lookup or create pipe
14-05-19 15:45:09 D:     50:1-pipe1             :         $ open pipe reader
14-05-19 15:45:09 1-pipe1: attach local reader
14-05-19 15:45:09 1-pipe1: broadcast we are now reader
14-05-19 15:45:09 D:     50:1-pipe1             : open local reader
14-05-19 15:45:09 D:     50:1-pipe1             :         > before reading
14-05-19 15:45:09 D:     50:1-pipe1             : before reading:
14-05-19 15:45:09 D:     50:1-pipe1             :     ok
14-05-19 15:45:09 D:     50:1-pipe1             :         $ send INPUT_OK
14-05-19 15:45:09 D:     50:1-pipe1             :         > reading
14-05-19 15:45:09 D:     50:1-pipe1             : reading:
14-05-19 15:45:09 D:     50:1-pipe1             :     READ
14-05-19 15:45:09 D:     50:1-pipe1             :         $ process read request
14-05-19 15:45:09 D:     50:1-pipe1             :         > processing read
14-05-19 15:45:12 D:     50:1-pipe1             : processing read:
14-05-19 15:45:12 D:     50:1-pipe1             :     wakeup
14-05-19 15:45:12 D:     50:1-pipe1             :         $ send READ_TIMEOUT
14-05-19 15:45:12 D:     50:1-pipe1             :         > reading
interrupted
14-05-19 15:45:13 N: terminating zpipes_server service
Context was terminated
Context was terminated
Context was terminated
Context was terminated

```

**Comment by hintjens** [comment]
There's something unusual with your network setup... hosts are receiving their own beacons, which is normally impossible. They're filtered out by network address. So perhaps hosts with multiple interfaces where there's some mapping happening that Zyre/zbeacon isn't designed to deal with...?

**Comment by rpedde** [comment]
These hosts are kind of crazy.  They have a ton of interfaces, and three of them are common across all hosts.  They are all rfc-1918 networks, and are vlan split.  One is a /24, two are /23.  They are all properly configured netmask-wise.

**Comment by rpedde** [comment]
This does appear to be a multiple-interface issue.  Hard-setting the ZSYS_INTERFACE environment  variable looks like it did the trick.

**Comment by hintjens** [comment]
Great... I'll make that a config file option then.

**Comment by hintjens** [comment]
Fixed by #81; zyre/interface config option specifies which network interface to use.
